extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/S3UploadManager.java

```diff
@@ -87,25 +102,36 @@ public Future<UploadPartResult> queueChunkForUpload(
       S3OutputConfig config
   )
   {
-    return uploadExecutor.submit(() -> RetryUtils.retry(
-        () -> {
-          log.debug("Uploading chunk[%d] for uploadId[%s].", chunkNumber, uploadId);
-          UploadPartResult uploadPartResult = uploadPartIfPossible(
-              s3Client,
-              uploadId,
-              config.getBucket(),
-              key,
-              chunkNumber,
-              chunkFile
-          );
-          if (!chunkFile.delete()) {
-            log.warn("Failed to delete chunk [%s]", chunkFile.getAbsolutePath());
-          }
-          return uploadPartResult;
-        },
-        S3Utils.S3RETRY,
-        config.getMaxRetry()
-    ));
+    final ServiceMetricEvent.Builder builder = new ServiceMetricEvent.Builder();
+    final Stopwatch taskQueuedStopwatch = Stopwatch.createStarted();
+    queueSize.incrementAndGet();
+    return uploadExecutor.submit(() -> {
+      emitter.emit(builder.setMetric(TASK_QUEUED_DURATION_METRIC, taskQueuedStopwatch.millisElapsed()));
+      emitter.emit(builder.setMetric(NUM_TASKS_QUEUED_METRIC, queueSize.decrementAndGet()));
+      final Stopwatch taskDurationStopwatch = Stopwatch.createStarted();
+      return RetryUtils.retry(
+          () -> {
+            log.debug("Uploading chunk[%d] for uploadId[%s].", chunkNumber, uploadId);
+            UploadPartResult uploadPartResult = uploadPartIfPossible(
+                s3Client,
+                uploadId,
+                config.getBucket(),
+                key,
+                chunkNumber,
+                chunkFile
+            );
+            if (!chunkFile.delete()) {
+              log.warn("Failed to delete chunk [%s]", chunkFile.getAbsolutePath());
+            }
+            emitter.emit(builder.setMetric(TASK_DURATION_METRIC, taskDurationStopwatch.millisElapsed())
+                                .setDimension("uploadId", uploadId)
+                                .setDimension("partNumber", chunkNumber));
```

Comment:
You can include these dimensions in the queued time metric as well. Also, the dimensions need to be set on the builder before calling `emitter.emit()`.